### PR TITLE
Add system status warning to display scheduled job failures

### PR DESCRIPTION
**Pull request by jitendrapurohit**
Overview
----------------------------------------
Add alert to display scheduled job errors

Before
----------------------------------------
Schedule job fails silently. Civi core doesn't provide an inbuilt option to notify admins.

From https://chat.civicrm.org/civicrm/pl/8wbq1y3j4tdeumnf1iuie44a5o

>anyone else wondered about having Sch Reminder errors listed as problems in System Status? Just found a case where 5 months back 'somehow' a membership was given an End Date of 0000:00:00 and guess what. that stops renewal reminders being sent out. sheesh.

After
----------------------------------------
Alert displayed on the system status page. Eg

image

### CODE REVIEW

**Comment by sunilpawar** [comment]
i have built similar functionality using extension : https://github.com/Skvare/com.skvare.jobchecker

File: CRM/Utils/Check/Component/Env.php [hunk 1]
```diff
@@ -688,6 +688,51 @@ public function checkExtensions() {
...
+      if (!empty($lastExecutionMessage) && strpos($lastExecutionMessage, 'Failure') !== FALSE) {
+        $viewLogURL = CRM_Utils_System::url('civicrm/admin/joblog', "jid={$job['id']}&reset=1");
+        $html .= "<tr><td>{$job['name']} </td><td>{$lastExecutionMessage}</td><td>{$job['last_run']}</td><td> <a href='{$viewLogURL}'>View Job Log</a></td></tr>";
+      }
+    }
+    if (empty($html)) {
+      return [];
+    }
+
+    $message = "<p>The following scheduled jobs failed on the last run:</p>
```
Comment (mlutfy) [inline CRM/Utils/Check/Component/Env.php:718]: Could you wrap the strings in `ts()` ? (this line, and other lines around it)

Comment (jitendrapurohit) [reply CRM/Utils/Check/Component/Env.php]: Sorry for the late update @mlutfy. I've added the ts() now. Thanks.

**Comment by mlutfy** [comment]
@jitendrapurohit Unfortunately it's not quite right - please see: https://docs.civicrm.org/dev/en/latest/translation/#best-practices

For example:

-  `$message = ts("<p>The following scheduled jobs failed on the last run:</p>`  - wrap just the string, not the HTML (when possible)
-  `ts($message),`  - do not include variables in `ts`  (unless it uses placeholders), but besides it is already translated a few lines above.

**Comment by jitendrapurohit** [comment]
Thanks @mlutfy I've updated the usage in the latest commit 👍

File: CRM/Utils/Check/Component/Env.php [hunk 2]
```diff
@@ -688,6 +688,54 @@ public function checkExtensions() {
...
+        'job_id' => $job['id'],
+        'options' => ['sort' => "id desc", 'limit' => 1],
+      ])['values'][0]['description'] ?? NULL;
+      if (!empty($lastExecutionMessage) && strpos($lastExecutionMessage, 'Failure') !== FALSE) {
+        $viewLogURL = CRM_Utils_System::url('civicrm/admin/joblog', "jid={$job['id']}&reset=1");
+        $html .= '<tr>
+          <td>' . ts('%1', [1 => $job['name']]) . ' </td>
+          <td>' . ts('%1', [1 => $lastExecutionMessage]) . '</td>
+          <td>' . ts('%1', [1 => $job['last_run']]) . '</td>
+          <td>' . ts('<a href=\'%1\'>View Job Log</a>', [1 => $viewLogURL]) . '</td>
```
Comment (mlutfy) [inline CRM/Utils/Check/Component/Env.php:715]: Can you do instead:

```
<td><a href="' . $viewLogURL . '"> . ts('View Job Log') . '</a></td>
```

This way it avoids adding a new string to translate (and HTML should be avoided when possible).

```
$ grep -r 'View Job Log'
CRM/Admin/Page/Job.php:          'name' => ts('View Job Log'),
```

File: CRM/Utils/Check/Component/Env.php [hunk 3]
```diff
@@ -688,6 +688,54 @@ public function checkExtensions() {
...
+      $lastExecutionMessage = civicrm_api3('JobLog', 'get', [
+        'sequential' => 1,
+        'return' => ["description"],
+        'job_id' => $job['id'],
+        'options' => ['sort' => "id desc", 'limit' => 1],
+      ])['values'][0]['description'] ?? NULL;
+      if (!empty($lastExecutionMessage) && strpos($lastExecutionMessage, 'Failure') !== FALSE) {
+        $viewLogURL = CRM_Utils_System::url('civicrm/admin/joblog', "jid={$job['id']}&reset=1");
+        $html .= '<tr>
+          <td>' . ts('%1', [1 => $job['name']]) . ' </td>
```
Comment (mlutfy) [inline CRM/Utils/Check/Component/Env.php:712]: `ts` is not necessary here. Job names are not translatable (they are set/translated during the installation, in `civicrm_data.xx_XX.mysql`).

File: CRM/Utils/Check/Component/Env.php (hunk 2 above)
Comment (jitendrapurohit) [reply CRM/Utils/Check/Component/Env.php]: Updated with the above suggestion 👍

File: CRM/Utils/Check/Component/Env.php (hunk 3 above)
Comment (jitendrapurohit) [reply CRM/Utils/Check/Component/Env.php]: Updated

File: CRM/Utils/Check/Component/Env.php [hunk 4]
```diff
@@ -688,6 +688,54 @@ public function checkExtensions() {
...
+          <td>' . ts('%1', [1 => $lastExecutionMessage]) . '</td>
+          <td>' . ts('%1', [1 => $job['last_run']]) . '</td>
```
[the remark this answers is not in the document]
Comment (jitendrapurohit) [reply CRM/Utils/Check/Component/Env.php]: Removed now. Thanks.

**Comment by mlutfy** [comment]
@jitendrapurohit Thank you!

Merging based on code review, feedback from other reviewers, and also `r-run`:

image

**Comment by demeritcowboy** [comment]
I remember seeing this PR and not really paying attention because I always dealt with this issue another way, but after rolling out 5.46 I see this is a good thing™. 🎉

**Pull request merged**